libs/api/ebpf_api.cpp

```diff
@@ -562,6 +720,84 @@ _update_map_element(
 }
 CATCH_NO_MEMORY_EBPF_RESULT
 
+static ebpf_result_t
+_update_map_element_batch(
+    ebpf_handle_t map_handle,
+    _In_opt_ const void* key,
+    size_t key_size,
+    _In_ const void* value,
+    size_t value_size,
+    _Inout_ uint32_t* count,
+    uint64_t flags) NO_EXCEPT_TRY
+{
+    EBPF_LOG_ENTRY();
+    ebpf_result_t result;
+    ebpf_protocol_buffer_t request_buffer;
+    ebpf_operation_map_update_element_batch_request_t* request;
+    ebpf_operation_map_update_element_batch_reply_t reply;
+    size_t input_count = *count;
+
+    ebpf_assert(value);
+    ebpf_assert(key || !key_size);
+
+    // Compute the maximum number of entries that can be updated in a single batch.
+    size_t max_entries_per_batch = 0;
+    max_entries_per_batch = UINT16_MAX - EBPF_OFFSET_OF(ebpf_operation_map_update_element_batch_request_t, data);
+    max_entries_per_batch /= (key_size + value_size);
+
+    try {
+        for (size_t key_index = 0; key_index < input_count;) {
+            // Compute the number of entries to update in this batch.
+            size_t entries_to_update = min(input_count - key_index, max_entries_per_batch);
+
+            request_buffer.resize(
+                EBPF_OFFSET_OF(ebpf_operation_map_update_element_batch_request_t, data) +
+                entries_to_update * (key_size + value_size));
+            request = reinterpret_cast<ebpf_operation_map_update_element_batch_request_t*>(request_buffer.data());
+
+            request->header.length = static_cast<uint16_t>(request_buffer.size());
+            request->header.id = ebpf_operation_id_t::EBPF_OPERATION_MAP_UPDATE_ELEMENT_BATCH;
+            request->handle = (uint64_t)map_handle;
+            request->option = static_cast<ebpf_map_option_t>(flags);
+
+            for (size_t index = 0; index < entries_to_update; index++) {
+                uint8_t* source_key = (uint8_t*)key + (key_index + index) * key_size;
+                uint8_t* source_value = (uint8_t*)value + (key_index + index) * value_size;
+                uint8_t* destination_key = request->data + index * (key_size + value_size);
+                uint8_t* destination_value = request->data + index * (key_size + value_size) + key_size;
+                if (key_size > 0) {
```

Comment:
Is a key_size of value 0 supported? Is a value_size of 0 supported as well?

(If both yes, then maybe a minor optimization can be taken  as this loop is a no-op, but then if that's the case, maybe we aren't even entering this loop at all?)

Reply:
It looks like key_size of 0 would be supported - in practice, when is that the case?

Reply:
Key_size can be zero for maps with no keys, examples are ringbuffer, stack, and queue.